README.md
# projlocator

Initial commit.
## On-line Documentation

More information on the **projlocator** project can be found
[here](https://jddixon.github.io/projlocator)


CHANGES
py/projlocator/CHANGES

v0.0.11
    2018-03-22
        * fix corrupt .dvcz/builds                                  SLOC  937
v0.0.10
    2018-03-06
        * edit setup.py, tox config files                           SLOC  933
v0.0.9
    2017-12-17
        * pylinting                                                 SLOC  939
v0.0.8
    2017-10-04
        * add .dvcz/projConfig.toml                                 SLOC  945
v0.0.7
    2017-09-16
        * correct tox.ini, requirements.txt                         SLOC  914
v0.0.6
    2017-08-17
        * rewrite setup.py; add tox.ini, *requirements.txt          SLOC  906
v0.0.5
    2017-07-22
        * add MANIFEST.in, rewrite setup.py; edit cfg files         SLOC  762
v0.0.4
    2017-06-17
        * restructure for pytest                                    SLOC  759
v0.0.3
    2017-04-29
        * correct handling of ruby and ocaml
        * drop dist/* from repos                                    SLOC  860
v0.0.2
    2017-01-27
        * pylinting; correct use of os.path.join                    SLOC  685
v0.0.1
    2016-11-30
        initial commit                                              SLOC  868













test_proj_list.py
#!/usr/bin/python3
# testProjList.py

""" Verify that the project list is correct and is read correctly. """

import unittest

from projlocator import PROJ_LIST_MAP, add_to_proj_list
# from rnglib import SimpleRNG


class TestProjList(unittest.TestCase):
    """ Verify that the project list is correct and is read correctly. """

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def check(self, project, rel_path):
        """ Verify the relative path for the project is as expected. """

        self.assertEqual(PROJ_LIST_MAP[project], rel_path)

    def test_proj_adder(self):
        """ Verify that the add_to_proj_list() funtion works correctly. """

        self.assertEqual(add_to_proj_list('foo_c'), 'c/foo_c')
        self.assertEqual(add_to_proj_list('foo_cpp'), 'cpp/foo_cpp')
        self.assertEqual(add_to_proj_list('foo_go'),
                         'go/src/github.com/jddixon/foo_go')
        self.assertEqual(add_to_proj_list('foo_java'), 'java/foo_java')
        self.assertEqual(add_to_proj_list('foo_ml'), 'ml/foo_ml')
        self.assertEqual(add_to_proj_list('foo_py'), 'py/foo_py')
        self.assertEqual(add_to_proj_list('foo_rb'), 'rb/foo_rb')

        try:
            # pylint: disable=no-value-for-parameter
            add_to_proj_list()                     # missing required parameter
            self.fail('succesfully added project without name')
        except TypeError:
            pass

        try:
            add_to_proj_list('foo')                # missing relPath
            self.fail('succesfully added project without relPath')
        except RuntimeError:
            pass

        try:
            add_to_proj_list('foo_py', 'foo.cpp')  # wrong relPath
            self.fail('succesfully added project without relPath')
        except RuntimeError:
            pass

        try:
            add_to_proj_list('pysloc', 'py')        # existing project
            self.fail('succesfully added existing project!')
        except RuntimeError:
            pass

    def test_proj_list_map(self):
        """ Verift that existing projects have the correct relative path. """

        self.check('alertz', 'py/alertz')
        self.check('gotwitgo', 'go/src/github.com/jddixon/gotwitgo')
        self.check('xlattice', 'xlattice')


if __name__ == '__main__':
    unittest.main()


test_dir_lang_maps.py
#!/usr/bin/python3
# testDirLangMaps.py

""" Verify functioning of get_lang_from_dir(). """

import time
import unittest

from rnglib import SimpleRNG
from projlocator import get_dir_for_lang, get_lang_for_dir


class TestMaps(unittest.TestCase):
    """ Verify functioning of get_lang_from_dir(). """

    def setUp(self):
        self.rng = SimpleRNG(time.time())

    def tearDown(self):
        pass

    def test_maps(self):
        """
        Actual test of get_lang_from_dirs(), which examines a path to
        determine the main language of a package.
        """

        # dir -> lang tests -----------------------------------------
        try:
            get_lang_for_dir(None)
            self.fail("didn't catch missing directory name")
        # pylint: disable=bare-except
        except:
            pass

        try:
            get_lang_for_dir('foo/bar')
            self.fail("didn't catch relative path")
        # pylint: disable=bare-except
        except:
            pass

        # failure to match just returns None
        self.assertEqual(get_lang_for_dir('/'), None)
        self.assertEqual(get_lang_for_dir('/foo'), None)

        self.assertEqual(
            get_lang_for_dir('/home/jdd/dev/gh-pages/projects/'), 'G')
        self.assertEqual(get_lang_for_dir('/home/jdd/dev/c/'), 'c')
        self.assertEqual(get_lang_for_dir('/home/jdd/dev/cpp/'), 'cpp')
        self.assertEqual(get_lang_for_dir('/home/jdd/dev/java/'), 'java')
        self.assertEqual(get_lang_for_dir('/home/jdd/dev/node/'), 'js')
        self.assertEqual(get_lang_for_dir('/home/jdd/dev/py/'), 'py')
        self.assertEqual(
            get_lang_for_dir('/home/jdd/dev/go/src/github.com/jddixon/'), 'go')
        self.assertEqual(get_lang_for_dir('/home/jdd/dev/ocaml/'), 'ml')
        self.assertEqual(get_lang_for_dir('/home/jdd/dev/ruby/'), 'rb')

        # dummy 'top' language
        self.assertEqual(get_lang_for_dir('/home/jdd/dev/'), 'top')

        # lang -> dir tests -----------------------------------------
        self.assertEqual(get_dir_for_lang('foo'), None)
        self.assertEqual(
            get_dir_for_lang('G'),
            '/home/jdd/dev/gh-pages/projects/')
        self.assertEqual(get_dir_for_lang('c'), '/home/jdd/dev/c/')
        self.assertEqual(get_dir_for_lang('cpp'), '/home/jdd/dev/cpp/')
        self.assertEqual(get_dir_for_lang('java'), '/home/jdd/dev/java/')
        self.assertEqual(get_dir_for_lang('js'), '/home/jdd/dev/node/')
        self.assertEqual(get_dir_for_lang('py'), '/home/jdd/dev/py/')
        self.assertEqual(get_dir_for_lang('go'),
                         '/home/jdd/dev/go/src/github.com/jddixon/')
        self.assertEqual(get_dir_for_lang('ml'), '/home/jdd/dev/ocaml/')
        self.assertEqual(get_dir_for_lang('rb'), '/home/jdd/dev/ruby/')

        # dummy 'top' language
        self.assertEqual(get_dir_for_lang('top'), '/home/jdd/dev/')

if __name__ == '__main__':
    unittest.main()


test_lang_for_projects.py
#!/usr/bin/python3
# sharedev/testLangForProject.py

""" Check functioning of get_lang_for_project. """

import time
import unittest

from rnglib import SimpleRNG
from projlocator import get_lang_for_project


class TestLangForProject(unittest.TestCase):
    """ Check functioning of get_lang_for_project. """

    def setUp(self):
        self.rng = SimpleRNG(time.time())

    def tearDown(self):
        pass

    # utility functions #############################################

    # actual unit tests #############################################
    def test_mapping(self):
        """
        Various tests of get_lang_for_project(), some with expected
        results, some expected to fail.
        """

        # dir -> lang tests -----------------------------------------
        try:
            get_lang_for_project(None)
            self.fail("didn't catch missing project name")
        # pylint: disable=bare-except
        except:
            pass

        # failure to match should return ""
        self.assertEqual(get_lang_for_project('/'), "")
        self.assertEqual(get_lang_for_project('foo'), "")

        # these names must be filtered out
        self.assertEqual(get_lang_for_project('dot'), "")
        self.assertEqual(get_lang_for_project('ghp.css'), "")
        self.assertEqual(get_lang_for_project('img'), "")
        self.assertEqual(get_lang_for_project('LICENSE.md'), "")
        self.assertEqual(get_lang_for_project('TODO'), "")

        # these are real project names
        self.assertEqual(get_lang_for_project('xlreg_c'), 'c')
        self.assertEqual(get_lang_for_project('xlreg_cpp'), 'cpp')
        self.assertEqual(get_lang_for_project('xlreg_java'), 'java')
        self.assertEqual(get_lang_for_project('cryptoserver_go'), 'go')
        self.assertEqual(get_lang_for_project('cryptoserver_go'), 'go')
        self.assertEqual(get_lang_for_project('ctries_go'), 'go')
        self.assertEqual(get_lang_for_project('gotwitgo'), 'go')
        self.assertEqual(get_lang_for_project('xgo_go'), 'go')
        self.assertEqual(get_lang_for_project('xlreg_ml'), 'ml')
        self.assertEqual(get_lang_for_project('merkletree'), 'py')
        self.assertEqual(get_lang_for_project('rnglib'), 'py')
        self.assertEqual(get_lang_for_project('xlreg_py'), 'py')
        self.assertEqual(get_lang_for_project('xlreg_rb'), 'rb')

        # top-level project(s)
        self.assertEqual(get_lang_for_project('xlattice'), 'top')

if __name__ == '__main__':
    unittest.main()


src/projlocator/__init__.py
# projlocator/projlocator/loc.py

""" Utilities for managing a projects directory organized by language. """

import os
import shutil

__all__ = ['__version__', '__version_date__',
           # CONSTANTS
           'DEV_BASE', 'SHAREDEV_APP_DIR',
           'GH_PAGES_BASE',
           'PROJ_LIST_MAP',
           # FUNCTIONS
           'add_to_proj_list', 'drop_from_proj_list',
           'get_dir_for_lang', 'get_gh_pages_base', 'get_gh_pages_proj',
           'get_lang_for_dir', 'get_lang_for_project',
           'get_proj_defaults', 'get_proj_names',
           'proj_dir_from_name', 'proj_rel_path_from_name',
           ]

# -- exported constants ---------------------------------------------

__version__ = '0.0.11'
__version_date__ = '2018-03-22'

DEV_BASE = os.path.join(os.environ['HOME'], 'dev')
SHAREDEV_APP_DIR = '/var/app/sharedev'
GH_PAGES_BASE = os.path.join(DEV_BASE, 'gh-pages', 'projects')

# -- pseudo constant(s) ---------------------------------------------
PROJ_LIST_FILE = os.path.join(SHAREDEV_APP_DIR, 'projList')
PROJ_LIST_MAP = {}


def _update_proj_list():
    # DEBUG
    # print("entering updateProjList()")
    # END
    global PROJ_LIST_MAP
    PROJ_LIST_MAP = {}
    with open(PROJ_LIST_FILE, 'r') as file:
        line = file.readline()
        while line:
            line = line.strip()
            if line:
                parts = line.split()
                PROJ_LIST_MAP[parts[0]] = parts[1]
            line = file.readline()


        # DEBUG
        # print("there are %d items in the map" % len(PROJ_LIST_MAP))
        # END
_update_proj_list()

# -- exported functions ---------------------------------------------


def add_to_proj_list(project, rel_path=None):
    """
    If the project does not already exist and if its relPath is not
    predictable from the project name, add it to the list in
    SHAREDEV_APP_DIR/projList.

    Raises a RuntimeError if a problem is encountered.  Otherwise
    returns the value of relPath used (ie, appended to projList).
    """
    if not project:
        raise RuntimeError('you must supply a project name!')

    if project in PROJ_LIST_MAP:
        raise RuntimeError("'%s' is already in the project list" % project)

    expected_rel_path = ''
    if project.endswith('_c'):
        expected_rel_path = os.path.join('c', project)
    elif project.endswith('_cpp'):
        expected_rel_path = os.path.join('cpp', project)
    elif project.endswith('_go'):
        expected_rel_path = os.path.join('go', 'src', 'github.com',
                                         'jddixon', project)
    elif project.endswith('_java'):
        expected_rel_path = os.path.join('java', project)
    elif project.endswith('_ml'):
        expected_rel_path = os.path.join('ml', project)
    elif project.endswith('_py'):
        expected_rel_path = os.path.join('py', project)
    elif project.endswith('_rb'):
        expected_rel_path = os.path.join('rb', project)

    if expected_rel_path:
        if rel_path:
            if rel_path != expected_rel_path:
                raise RuntimeError("relPath is '%s' but expected '%s'" % (
                    rel_path, expected_rel_path))
        else:
            rel_path = expected_rel_path
        # regular project name, so we don't update the relPath file
        return rel_path

    if not rel_path:
        raise RuntimeError(
            "'%s' is not a regular project name; you must specify a relPath" %
            project)

    tmp_file = PROJ_LIST_FILE + '.tmp'
    bak_file = PROJ_LIST_FILE + '.bak'
    shutil.copyfile(PROJ_LIST_FILE, tmp_file)
    with open(tmp_file, 'a') as file:
        file.write(project + '\t' + rel_path + '\n')

    # quasi-atomic file update
    os.rename(PROJ_LIST_FILE, bak_file)
    os.rename(tmp_file, PROJ_LIST_FILE)

    _update_proj_list()

    return rel_path


def drop_from_proj_list(project):
    """ Drop the project named from the project list. """

    if not project:
        return
    if project in PROJ_LIST_MAP:
        # DEBUG
        print("dropping project '%s'" % project)
        # END
        tmp_file = PROJ_LIST_FILE + '.tmp'
        bak_file = PROJ_LIST_FILE + '.bak'

        with open(PROJ_LIST_FILE, 'r') as file:
            data_in = file.read()
        lines = data_in.split('\n')
        if not lines[-1]:
            lines = lines[0:-1]
        output = []
        for line in lines:
            proj, _ = line.split()     # _ was r
            if proj != project:
                output.append(line)
        data_out = '\n'.join(output) + '\n'
        # DEBUG
        print("updated projList:")
        print(data_out)              # prints an extra line :-}
        # END
        with open(tmp_file, 'w') as file:
            file.write(data_out)

        # quasi-atomic file update
        os.rename(PROJ_LIST_FILE, bak_file)
        os.rename(tmp_file, PROJ_LIST_FILE)

        _update_proj_list()
    else:
        print("the project you are trying to drop, '%s', does not exist" %
              project)


def get_gh_pages_base():
    """ Return path to gh-pages subdirectory. """
    return os.path.join(os.environ['DEV_BASE'], 'gh-pages')


def get_gh_pages_proj():
    """ Return path to gh-pages/projects subdirectory. """
    return os.path.join(get_gh_pages_base(), 'projects')


def get_proj_names():
    """ get a sorted list of all projects in gh-pages"""

    names = []
    files = os.listdir(GH_PAGES_BASE)
    if files:
        files.sort()
        for file in files:
            proj_dir = proj_dir_from_name(file)
            if proj_dir == "":
                continue
            names.append(file)
    return names


def proj_dir_from_name(project):
    """ get path to project directory from name """
    proj_dir = ''
    if project == "testing":
        proj_dir = "/home/jdd/dev/py/wrap_gh_page"
    else:
        rel_path = proj_rel_path_from_name(project)
        if rel_path:
            proj_dir = os.path.join(DEV_BASE, rel_path)
    return proj_dir


def proj_rel_path_from_name(project):
    """
    Get relative path from base directory (dev/ or sharedev/) to
    project directory.
    """
    if not project:
        raise "no project name specified"

    rel_path = ""

    # IRREGULAR PROJECT NAMES ---------------------------------------
    if project in PROJ_LIST_MAP:
        rel_path = PROJ_LIST_MAP[project]

    # PROJECT NAMES WITH STANDARD SUFFIXES --------------------------
    # If you update this, update addToProjList() as well
    elif project.endswith('_c'):
        rel_path = os.path.join('c', project)
    elif project.endswith('_cpp'):
        rel_path = os.path.join('cpp', project)
    elif project.endswith('_data'):
        rel_path = os.path.join('dat', project)
    elif project.endswith('_go'):
        rel_path = os.path.join('go', 'src', 'github.com', 'jddixon', project)
    elif project.endswith('_java'):
        rel_path = os.path.join('java', project)
    elif project.endswith('_ml'):
        rel_path = os.path.join('ml', project)
    elif project.endswith('_py'):
        rel_path = os.path.join('py', project)
    elif project.endswith('_rb'):
        rel_path = os.path.join('rb', project)
    return rel_path


# -- lang/dir pairs -------------------------------------------------
PAIRS = [\
    # to simplify searching, all paths should be slash-terminated
    ['c', os.path.join(DEV_BASE, 'c/')],
    ['cpp', os.path.join(DEV_BASE, 'cpp/')],
    ['dat', os.path.join(DEV_BASE, 'dat/')],
    ['G', os.path.join(DEV_BASE, 'gh-pages', 'projects/')],
    ['go', os.path.join(DEV_BASE, 'go', 'src', 'github.com', 'jddixon/')],
    ['java', os.path.join(DEV_BASE, 'java/')],
    ['js', os.path.join(DEV_BASE, 'node/')],
    ['py', os.path.join(DEV_BASE, 'py/')],
    ['ml', os.path.join(DEV_BASE, 'ml/')],
    ['rb', os.path.join(DEV_BASE, 'rb/')],

    # dummy language
    ['top', DEV_BASE + '/'], ]
DIR2LANG = {}
LANG2DIR = {}
for pair_ in PAIRS:
    lang_ = pair_[0]
    dir_ = pair_[1]
    DIR2LANG[dir_] = lang_
    LANG2DIR[lang_] = dir_

# -- utility functions ----------------------------------------------


def get_lang_for_project(project):
    """
    Given a project name, return the language.

    ridiculously inefficient"""

    lang = ''
    abs_path = proj_dir_from_name(project)
    if abs_path:
        lang, _, _ = get_proj_defaults(abs_path)     # was _p, _r
    return lang


def get_proj_defaults(abs_path=None):
    """
    Given the absolute path , or the current directory if no path is
    specified, return the language code, project name, and relative path
    from the project directory.  If these cannot be resolved, return None
    in all cases.
    """

    if not abs_path:
        abs_path = os.getcwd()
    lang, project, rel_path = None, None, None
    failed = True
    for pair in PAIRS:
        lang_dir = pair[1]
        if abs_path.startswith(lang_dir):
            lang = pair[0]
            failed = False
            break
    if not failed:
        len_lang_dir = len(lang_dir)
        rest_of_path = abs_path[len_lang_dir:]
        project, beep, rel_path = rest_of_path.partition('/')
        if beep != '/':
            rel_path = './'

    if failed:
        return None, None, None
    return lang, project, rel_path


def get_dir_for_lang(lang):
    """
    Given a language code, returns the corresponding absolute path without
    any trailing slash, or None.
    """
    try:
        dir_name = LANG2DIR[lang]
    except KeyError:
        dir_name = None
    return dir_name


def get_lang_for_dir(dir_name):
    """
    Given an absolute path, returns the corresponding language code or
    None if there isn't one.
    """
    if dir_name is None:
        raise RuntimeError('path may not be empty')
    if not dir_name.startswith('/'):
        raise RuntimeError('path must be absolute')
    if not dir_name.endswith('/'):
        dir_name += '/'
    try:
        lang = DIR2LANG[dir_name]
    except KeyError:
        lang = None
    return lang
